### PR TITLE
Single kafka image

### PR DESCRIPTION
**Pull request by tombentley**
### Type of change

- Refactoring

### Description

This is an initial attempt to reduce the number of Kafka-based images which Strimzi uses.

Essentially what I've done here is to put the different scripts which were the only real difference between the numerous old images into a singe image. However, note that there are numerous details to review:

* The `Dockerfile` could probably do with tidying up.
* Against best-practice I no longer expose any ports in the `Dockerfile`, because the set of ports in common across all the users of the image is the empty set. 
* Because we weren't using the "image map" mechanism in the CO for configuring different ZK images depending on the Kafka version, I'm using the most recent Kafka version (or more correctly the last entry in the `kafka-versions` file) to supply the image used for ZK. This shouldn't matter as all current and foreseeable Kafka versions use the same version of ZK.
* So far I've tested this by hand the ZK, Kafka, KafkaConnectS2I images. I've not yet tested plain Connect or MirrorMaker
* I wondered what to do about the test-clients image. I felt we should not include in the production image scripts which we're only really intending to be used during testing. So currently this is still a separate image.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: kafka-versions [hunk 1]
```diff
@@ -22,7 +22,7 @@
 # version default  proto   msg   sha512
 #1.0.2              1.0     1.0   4CBCDAF8CCC4EFE3D1B6275F3F2C32CF8F2F1A62104B5DD0BD9E2974160AB89D85A6E1791AF8B948A413B99ED696B06EA9D4299B27EA63C3F7318DABF5761144
 #1.1.1              1.1     1.1   2A1EB9A7C8C8337C424EEFED7BAAE26B3DACBA6A4AB8B64D9A7D5C6EE2CDB66CFA76C5B366F23435941569B89BF02482625189016296B2EA2A05FD0F38F6B709
-2.0.0              2.0     2.0   b28e81705e30528f1abb6766e22dfe9dae50b1e1e93330c880928ff7a08e6b38ee71cbfc96ec14369b2dfd24293938702cab422173c8e01955a9d1746ae43f98
+#2.0.0              2.0     2.0   b28e81705e30528f1abb6766e22dfe9dae50b1e1e93330c880928ff7a08e6b38ee71cbfc96ec14369b2dfd24293938702cab422173c8e01955a9d1746ae43f98
```
Comment (scholzj) [inline kafka-versions]: Any particular reason for this? Or is it just a leftover from some test?

File: docker-images/kafka/Dockerfile [hunk 2]
```diff
@@ -1,11 +1,61 @@
...
+# exposing Kafka ports and the one for JMX exporter
+#EXPOSE 9091 9092 9404
+
+
+
+
 USER 1001
 
-CMD ["/opt/kafka/kafka_run.sh"]
+#CMD ["/opt/kafka/kafka_run.sh"]
```
Comment (scholzj) [inline docker-images/kafka/Dockerfile]: Should we just deleted the commented parts?

**Comment by kyguy** [comment]
Looks good to me, just have a couple of noob questions/comments

1. ) I take it that we can expose the container ports using the cluster operator itself? Is that why they can be excluded from the Docker file?
2. ) The inclusion/exclusion of the test client doesn't matter to me since it has such a small footprint, I support whatever is easiest for everyone

File: docker-images/kafka/Dockerfile [hunk 3]
```diff
@@ -1,11 +1,61 @@
...
+# Set S2I folder
+ENV S2I_HOME=/opt/kafka/s2i
+
+# Copy S2I scripts
+COPY ./s2i-scripts $S2I_HOME
+
+WORKDIR $KAFKA_HOME
+
+# exposing Kafka ports and the one for JMX exporter
+#EXPOSE 9091 9092 9404
```
Comment (ppatierno) [inline docker-images/kafka/Dockerfile]: comment to delete?

File: docker-images/kafka/Dockerfile [hunk 4]
```diff
@@ -1,11 +1,61 @@
...
+# Downloading/extracting Apache Kafka
+RUN curl -O https://www.apache.org/dist/kafka/${KAFKA_VERSION}/kafka_${SCALA_VERSION}-${KAFKA_VERSION}.tgz \
+    && echo $KAFKA_CHECKSUM > kafka_${SCALA_VERSION}-${KAFKA_VERSION}.tgz.sha512 \
+    && sha512sum --check kafka_${SCALA_VERSION}-${KAFKA_VERSION}.tgz.sha512 \
+    && mkdir $KAFKA_HOME \
+    && tar xvfz kafka_${SCALA_VERSION}-${KAFKA_VERSION}.tgz -C $KAFKA_HOME --strip-components=1 \
+    && rm -f kafka_${SCALA_VERSION}-${KAFKA_VERSION}.tgz*
 
 # copy scripts for starting Kafka
 COPY ./scripts/ $KAFKA_HOME
```
Comment (ppatierno) [inline docker-images/kafka/Dockerfile]: I see that all scripts will go together in the `KAFKA_HOME`. Just for tidying it up a little bit, could we have subfolders for `kafka`, `zookeeper`, `connect`, etc etc ... and `utility` (or `common`) scripts?

**Comment by ppatierno** [comment]
LGTM overall, I agree on the test client image as well.

File: kafka-versions (hunk 1 above)
Comment (tombentley) [reply kafka-versions]: I'll remove it, but the question is, in 0.12.x are we going to continue providing 2.0.0 when 0.11.x provides 2.0.1? There's little cost to having it, but it does slow down builds.

File: docker-images/kafka/Dockerfile (hunk 4 above)
Comment (tombentley) [reply docker-images/kafka/Dockerfile]: TBH there's a bit of work in doing and testing that and for very little benefit that I can see. I'll consider doing it an another PR.

**Comment by tombentley** [comment]
> 1) I take it that we can expose the container ports using the cluster operator itself? Is that why they can be excluded from the Docker file?

It's best practice to use `EXPOSE`, but my understanding is that it's really only providing metadata. See https://docs.docker.com/engine/reference/builder/#expose. The container ports in a kube container overrides the exposed ports anyway. Since we're both the producer and consumer of this image it's documentary value to us is of basically no value.

> 2) The inclusion/exclusion of the test client doesn't matter to me since it has such a small footprint, I support whatever is easiest for everyone.

The test client is only needed for the system tests. It's not something we intend end users to use (even indirectly).

File: docker-images/kafka/Dockerfile (hunk 4 above)
Comment (ppatierno) [reply docker-images/kafka/Dockerfile]: I know that the benefit is not great, it's about organizing things and not having all different stuff in one place. Btw we can leave with the chaos :-)

File: kafka-versions (hunk 1 above)
Comment (scholzj) [reply kafka-versions]: I think we need to have some strategy for this which we will follow. We can't keep adding releases forever. 2.2.0 will be released soon. So I think we need to talk about things like when are we going to drop 2.0.X completely etc. Anyway, that is a bit beyond the scope of this PR I guess.

**Pull request merged**